src/fondant/pipeline/pipeline.py

```diff
@@ -362,16 +387,43 @@ def read(
                 msg,
             )
 
-        operation = ComponentOp(
-            name_or_path,
-            produces=produces,
-            arguments=arguments,
-            input_partition_rows=input_partition_rows,
-            resources=resources,
-            cache=cache,
-            cluster_type=cluster_type,
-            client_kwargs=client_kwargs,
-        )
+        if inspect.isclass(ref) and issubclass(ref, PythonComponent):
+            name = ref.__name__
+            image = ref.image()
+            description = ref.__doc__ or "python component"
+
+            component_spec = ComponentSpec(
+                name,
+                image.base_image,  # TODO: revisit
+                description=description,
+                consumes={"additionalProperties": True},
+                produces={"additionalProperties": True},
+            )
+
+            operation = ComponentOp(
+                name,
+                image,
+                component_spec,
+                produces=produces,
+                arguments=arguments,
+                input_partition_rows=input_partition_rows,
+                resources=resources,
+                cache=cache,
+                cluster_type=cluster_type,
+                client_kwargs=client_kwargs,
+            )
+
+        else:
+            operation = ComponentOp.from_component_yaml(
+                ref,
+                produces=produces,
+                arguments=arguments,
+                input_partition_rows=input_partition_rows,
+                resources=resources,
+                cache=cache,
+                cluster_type=cluster_type,
+                client_kwargs=client_kwargs,
+            )
```

Comment:
This code is almost identically repeated three times. Would be good if we can abstract this away.